충남대 Android_전영주 5주차 과제 STEP 1 (피드백 반영)

aengzu opened the pull request:
### STEP1 피드백 반영
---
STEP 1 피드백 반영하였습니다.
- Binds 를 사용하여 UseCase와 Repo 의 구현체와 인터페이스를 bind 하였습니다. 
- 폴더 구조에서 data 레이어의 use case 관련 코드를 domain 으로 이동시켰습니다
- HttpService 관련 코드도 의존성 주입을 사용하였습니다. 
- 코루틴 디스패처를 사용하여 콜백 관련 코드를 리팩토링 하였습니다.
- @Qualifier 을 사용하여 Database 이름 상수를 provide 하였는데 이 부분은 아직 제대로 모르는 채로 진행하여 더 알아보겠습니다.

상세한 리뷰 주셔서 감사합니다 ☺️ !!

### 어려웠던 점 & 리뷰 받고 싶은 점
---
- 원래 기능별로 커밋을 하려 했지만 에러를 수정하는 과정에서 한 번에 커밋하게 된 것 같습니다. ㅠ 다음 부분이 1단계 요구사항 작성 부분입니다! [refactor: refactor to use Room and Hilt for DI](https://github.com/kakao-tech-campus-2nd-step2/android-map-refactoring/pull/30/commits/02ba4c0d9c488c4782be1149d79a6f764a81543d)
- di 폴더 내부의 Module 들을 처음 작성해봤는데 올바르게 작성되었는지 코드 리뷰를 받고 싶습니다! [app/src/main/java/campus/tech/kakao/map/di/DatabaseModule.kt](https://github.com/kakao-tech-campus-2nd-step2/android-map-refactoring/pull/30/commits/02ba4c0d9c488c4782be1149d79a6f764a81543d#diff-324065a7439e7260fa41d67505977e7827413b03055b777512eca9222d53ff79)
- 하다보니 제가 봤을 땐 다음 단계 미션까지 구현된 것 같은데 맞을까요? 2단계 미션은 'MVVM 아키텍처 패턴을 적용한다, DataBinding, LivaData를 사용한다, 비동기 처리를 Coroutine으로 변경한다' 이고 3단계 미션은 'Repository 패턴을 적용한다' 입니다. 만약 구현되지 않았다면 다음 단계에서 해야할 일이 무엇일지 알려주시면 감사하겠습니다!
- 초기에는 검색어 목록에 대한 Entity 로 Primary key 를 자동으로 증가하는 id 로 설정을 했더니 같은 검색 히스토리를 중복하여 저장할 때 중복 값이 들어가길래 Primary Key 을 쿼리 String 그 자체로 하여 중복 값을 제외시켰습니다. 다음처럼 기본키를 placeName 으로 설정해도 될까요?  [fix: update search history order and prevent duplicate entries](https://github.com/kakao-tech-campus-2nd-step2/android-map-refactoring/pull/30/commits/456752383808f1713df40ea478956463594d9dad)
- 마지막 장소를 저장할 때 들어가는 원소는 하나인데 last_place 테이블을 따로 만드는 게 맞는지 잘모르겠어서 visited_places 테이블을 만들어서 방문한 모든 장소를 저장하고 마지막 장소에 접근할 때는 가장 id 가 높은(가장 최근에 저장된) 데이터를 DAO 로 접근하였습니다. 그런데 지금 생각해보니 불필요하게 데이터가 계속해서 저장되는 것 같은데 더 좋은 방법이 있을까요? (data/local/db/VIsitedPlaceDao.kt) [app/src/main/java/campus/tech/kakao/map/data/local/entity/VisitedPlaceEntity.kt](https://github.com/kakao-tech-campus-2nd-step2/android-map-refactoring/pull/30/commits/02ba4c0d9c488c4782be1149d79a6f764a81543d#diff-0627bd10713b0993384e7c3f9714b30467b3b4979f03c4e4b80518017f50467c), [app/src/main/java/campus/tech/kakao/map/data/local/db/VisitedPlaceDao.kt](https://github.com/kakao-tech-campus-2nd-step2/android-map-refactoring/pull/30/commits/02ba4c0d9c488c4782be1149d79a6f764a81543d#diff-3e5573f3631729d4d20ac2011b67152d37ce435464ab3650888bdd7bcd7915c2)
- 원격으로 받아오는 데이터(kakao local api)와 어플에서 사용하는 데이터를 모두 관리하다보니 폴더 구조를 어떻게 해야할 지 정하는 부분이 어려웠습니다. Room 을 사용하게 되어 SharePref 를 사용했던 기존 DBHelper 클래스는 제거하고 data, domain, presentation 중 data 레이어에서 remote 와 local 로 나누어 local 안에 관련 코드를 넣었습니다. local 안에 db 와 entity 폴더를 만들어서 검색 히스토리와 마지막 장소에 대한 entity 코드들과 DAO 코드들을 넣었습니다. 그러다보니 remote/entity 와 local/entity 가 혼동이 와서 인터넷에 검색했는데 데이터를 전달할 때 처럼 일회성으로 쓰이는 건 DTO 라는 용어를 많이 쓴다는 것을 알게되어  remote/entity 를 remote/dto 로 변경하였습니다. (아래 사진이 폴더 구조) 그런데 아직 remote/dto 내부에 있는 파일 명들 (DocumentEntity.kt, MetaEntity.kt, SearchResponse.kt) 는 따로 변경하지 않았습니다. 내부 파일 명들도 ㅇㅇDto.kt 로 변경하는 것이 좋을까요? 또한 현재 폴더 구조에서 개선점이 알고 싶습니다. 
<img width="339" alt="스크린샷 2024-07-24 오전 2 09 35" src="https://github.com/user-attachments/assets/7c319aab-1bfb-4abf-85b9-66b6ae21bbfc">

MyStoryG committed:
Initial commit

aengzu committed:
feat: step0 (#17)

aengzu committed:
충남대Android_전영주_4주차 과제_Step1 (피드백 반영) (#37)

* feat: step0

* docs: 1단계 기능 목록 작성

* feat(ui): create BottomSheet layout

* feat: 마커 기능&바텀 시트 추가

* feat: 에러 화면 추가

* feat: 마지막 위치를 저장 및 앱 실행 시 포커스

* feat: 피드백 반영

aengzu committed:
충남대 Android_전영주 4주차 Step2 (#83)

* feat: step0

* docs: 1단계 기능 목록 작성

* feat(ui): create BottomSheet layout

* feat: 마커 기능&바텀 시트 추가

* feat: 에러 화면 추가

* feat: 마지막 위치를 저장 및 앱 실행 시 포커스

* feat: 피드백 반영

* feat: step2 UI테스트 코드 추가

aengzu committed:
Add 'week04/' from commit 'fb1d5d4152d928f975a3f449af5f4f661bb03ec6'

git-subtree-dir: week04
git-subtree-mainline: 305453d2154dc74b2392e8f73d3bfc629c0fe38f
git-subtree-split: fb1d5d4152d928f975a3f449af5f4f661bb03ec6

aengzu committed:
step0: 4주차 코드 가져오기

aengzu committed:
feat: add Room entity annotation to DocumentEntity

aengzu committed:
feat: add SearchQueryEntity

aengzu committed:
refactor: restructure project folders

aengzu committed:
refactor: refactor to use Room and Hilt for DI

- SearchQueryDao, VisitedPlaceDao 및 SearchQueryEntity, VisitedPlaceEntity 를
생성
- SharedPref 대신 Room 데이터베이스를 사용하도록 함
- 프로젝트 전반에 Hilt를 구현
- Hilt 사용을 위해 의존성 제공 모듈인 DatabaseModule, RepositoryModule,
	UsecaseModule 생성
- 메인 스레드 차단 방지를 위해 DB 작업은 백그라운드로 이동
- 원래 기능 별로 커밋하려 했으나 에러를 고치는 과정에서 다 연관되어 있어 한
번에 커밋합니다. ㅠ

aengzu committed:
fix: fix RecyclerView item click listener error

- 아이템 클릭 리스너가 작동하지 않은 문제 해결
- 원인: XML 레이아웃 파일에 clickListener 바인딩이 누락됨

aengzu committed:
fix: update search history order and prevent duplicate entries

- 검색 기록 목록의 순서를 최신 검색이 가장 왼쪽에 오도록
변경했습니다.(reversed 사용)
- 중복된 검색 기록이 추가되지 않도록 수정했습니다.

aengzu committed:
save the project

aengzu committed:
Merge branch 'aengzu' into step1

aengzu committed:
chore: recommit to correct previous commit

InyoungAum commented on app/src/main/res/layout/activity_map.xml:
```diff
@@ -75,7 +75,7 @@
                 android:layout_height="wrap_content"
                 android:layout_marginLeft="10dp"
                 android:layout_marginBottom="10dp"
-                android:text="@{viewModel.lastPlace.placeName}"
```
이 부분은 databinding이 제거 된 것 같은데 왜 바꾸신건가요?

aengzu replied:
처음에 에러가 떠서 뭐가 문제인지 몰라서 이것저것 고치다가 저렇게 된 것 같습니다. 피드백 반영할 때 수정하겠습니다!

InyoungAum commented on app/src/main/java/campus/tech/kakao/map/di/RepositoryModule.kt:
```diff
@@ -0,0 +1,23 @@
+    @Provides
+    @Singleton
+    fun provideRepository(searchQueryDao: SearchQueryDao, visitedPlaceDao: VisitedPlaceDao) : PlaceRepository {
+        return PlaceRepositoryImpl(searchQueryDao, visitedPlaceDao)
```
이렇게 Repo자체를 생성해도 가능하지만 이렇게 되면 Repo에서 굳이 inject를 해줄 필요가 없어집니다.
또한 완벽한 클린아키텍쳐 기반의 멀티모듈을 적용했을때 Repo의 구현체를 알고있어야하는 문제가 생기게 되는데요

https://hungseong.tistory.com/29

위의 블로그를 참고해서 Repo interface, RepoImpl을 생성하고 bind하는 방법으로 한번 변경해보시기 바랍니다

aengzu replied:
넵!!

InyoungAum commented on app/src/main/java/campus/tech/kakao/map/di/UsecaseModule.kt:
```diff
@@ -0,0 +1,73 @@
+
+@InstallIn(SingletonComponent::class)
+@Module
+object UsecaseModule {
```
Usecase들도 위의 바인드방법을 사용해서 바꿔보시길 바랍니다!

aengzu replied:
넵 감사합니다!

InyoungAum commented on app/src/main/java/campus/tech/kakao/map/di/DatabaseModule.kt:
```diff
@@ -0,0 +1,31 @@
+        return Room.databaseBuilder(
+            appContext,
+            AppDatabase::class.java,
+            "kakao_map"
```
이러한 고정된 String들도 주입하는 방법으로 쓰면 유용한데요
하나 이상의 같은 형태를 주입하기 위해서는 @qualifier 어노테이션을 사용해야합니다.

https://seosh817.tistory.com/76

위 블로그를 참고하여 String 상수들도 주입해보시기 바랍니다

aengzu replied:
알겠습니다~!

InyoungAum commented on app/src/main/java/campus/tech/kakao/map/data/repository/PlaceRepositoryImpl.kt:
```diff
@@ -1,61 +1,75 @@
-                } else {
-                    Log.d("testt", "Response failed: ${response.toString()}")
-                    callback(null)
+        HttpService.searchKeyword(query = query) { response ->
```
HttpService도 주입을 받으면 좋을 것 같네요!

aengzu replied:
넵!! 감사합니다

InyoungAum commented on app/src/main/java/campus/tech/kakao/map/data/repository/PlaceRepositoryImpl.kt:
```diff
@@ -1,61 +1,75 @@
+    private val searchQueryDao: SearchQueryDao,
+    private val visitedPlaceDao: VisitedPlaceDao
+) : PlaceRepository {
     override fun searchPlaces(query: String, callback: (List<PlaceVO>?) -> Unit) {
```
다음단게에 대한 질문을 해주셨는데요 
coroutine 사용의 핵심중 하나는 callback이 없는 선형적인 흐름이라고 생각합니다. 
따라서 이 부분을 코루틴을 사용하여 callback 없이 리턴을 받을 수 있게 변경하는 방법을 생각해보시면 좋을 것 같습니다!

aengzu replied:
아 그렇군요 넵 감사합니다!

InyoungAum reviewed (changes requested):
수고하셨습니다!

InyoungAum commented:
## 다음처럼 기본키를 placeName 으로 설정해도 될까요?

해당 테이블에서 placeName이 유일한 값으로 사용되어야 한다면 맞는 방법입니다. 
아마 저도 그렇게 작업헀을 것 같네요!

InyoungAum commented:
## ... 그런데 지금 생각해보니 불필요하게 데이터가 계속해서 저장되는 것 같은데 더 좋은 방법이 있을까요? 

데이터를 저장할 때 로컬에서 queryTime등의 필드를 추가해서 System.currentTimeinMill 등의 함수로 검색한 시간을 저장하고 
가장 큰값을 가져오는 형태로는 해결이 안될까요?

InyoungAum commented:
## 내부 파일 명들도 ㅇㅇDto.kt 로 변경하는 것이 좋을까요? 또한 현재 폴더 구조에서 개선점이 알고 싶습니다.

이전에 말씀드렸듯 몇몇 네이밍 룰이 있고 이는 팀에서 정하기 나름이라 구분하기 편하신 형태로 가져가면 될 것 같습니다. 
내부에서 데이터를 전달하는 클래스들을 모두 Entity로 한다던지 혹은 외부에서 받는 것은 Entity 로컬 디비에서만 사용하는건 DTO등으로 바꾸셔도 무방할 것 같습니다. 

지금 데이터 패키지 하위에 repo와 usecase가 모두 있는데요 여기서 멀티 모듈을 배우진 않으시겠지만 

data 모듈 하위에는 네트워크 통신 관련 + repository의 구현체 (oooRepositoryImpl) 등이 들어가게 됩니다. 
domain 모듈 하위에는 모든 use case와 구현체 + repository의 interface가 있도록 변경해보시기 바랍니다

aengzu committed:
fix: 데이터 바인딩 관련 피드백 반영

aengzu committed:
refactor: Update DI configuration for PlaceRepository and UseCases using Hilt

- 기존에는 Provide 를 사용여서 의존성 주입을 하였으나 interface, Impl 을 @Binds
를 통해 bind 하는 방법으로 변경했습니다
- 비동기 작업 처리를 위해 repo Impl 에 디스패처 제공하였는데 object
CoroutineDispatcherModule이 없으면 순환 참조 에러가 발생하여 이를
추가하였습니다. 이에 대한 추가적 공부를 주말에 해보겠습니다.

aengzu committed:
refactor: restructure project folders

- data 패키지 내의 usecase 구현체를 domain 모듈로 이동했습니다
- API 서버로 부터 받게 되는 폴더를 entity 로 변경하고 로컬에서 데이터 전달 시
사용되는 domain 모듈 내의 model 을 dto 로 변경하였습니다

aengzu committed:
refactor: add Qualifier for Database Name String injection

- 데이터베이스 문자열 주입을 위해 @Qualifier 을 사용하였습니다
- 아직 잘 모르는 개념이라 시간이 부족하여 주말동안 더 조사해보겠습니다.

aengzu committed:
refactor: inject HttpService using Hilt

- HttpService 도 의존성 주입을 사용하였습니다.
- APIKey 와 BaseUrl 을 네트워크 모듈내에서 제공하여 HttpService 를 provide
하였습니다.

aengzu committed:
save the project

InyoungAum reviewed (approved):
LGTM 
코드가 훨씬 나아진 것 같네요 고생 많으셨습니다!
